CIP-0PRS/README.lagda.md

```diff
@@ -0,0 +1,1503 @@
+| Healing time            | $T_\text{heal}$ | slots   | Healing period to mitigate a strong (25-50%) adversary.                                   | $T_\text{heal} = \mathcal{O}\left( B / f \right)$        | Sufficient blocks must be produced to overcome an adversarially boosted block.               |
+| Chain-quality time      | $T_\text{CQ}$   | slots   | Ensure the presence of at least one honest block on the chain.                            | $T_\text{CQ} = \mathcal{O} (k/f)$                        | A least one honest block must be produced.                                                   |
+| Common-prefix time      | $T_\text{CP}$   | slots   | Achieve settlement.                                                                       | $T_\text{CP} = \mathcal{O} (k/f)$                        | The Ouroboros Praos security parameter defines the time for having a common prefix.          |
+| Security parameter      | $k$             | blocks  | The Ouroboros Praos security parameter.                                                   | n/a                                                      | Value for the Cardano mainnet.                                                               |
+
+
+### Specification of votes and certificates
+
+The stake-proportional voting in Peras mimics the _sortition_ algorithm used in Praos: specifically it is based on the use of a *verifiable random function* (VRF) by each stake-pool operator guaranteeing the following properties:
+
+- The probability for each voter to cast their vote in a given round is correlated to their share of total stake.
+- It should be computationally impossible to predict a given SPO's schedule without access to their secret key VRF key.
+- Verification of a voter's right to vote in a round should be efficiently computable.
+- A vote should be unique and non-malleable, which is a requirement for the use of efficient certificates aggregation.
+
+Additionally one would like the following property to be provided by the voting scheme:
+
+- Voting should require minimal additional configuration (e.g., key management) for SPOs.
+- Voting and certificate construction should be fast in order to ensure it does not interfere with other operations happening in the node.
+
+The precise scheme and format for votes and certificates is immaterial to the protocol itself and is deferred to another proposed CIP [*Votes & Certificates on Cardano*](https://github.com/cardano-foundation/CIPs/pull/870).
+
+### CDDL schema for the ledger
+
+Peras requires a single addition, `peras_cert`, the [block](#blocks) data on the ledger.
+
+```diff
+ block =
+   [ header
+   , transaction_bodies         : [* transaction_body]
+   , transaction_witness_sets   : [* transaction_witness_set]
+   , auxiliary_data_set         : {* transaction_index => auxiliary_data }
+   , invalid_transactions       : [* transaction_index ]
++  , ? peras_cert               : alba_certificate
```

Comment:
The occasional addition of a Peras certificate with size of the order of the max block size will impact the diffusion of these blocks. The [section](https://github.com/cardano-scaling/CIPs/blob/peras/CIP-0PRS/README.lagda.md#resource-requirements) on resource requirements focuses on localized factors, what about the emergent global impact?

I can imagine that slot leaders that produce these blocks, might diffuse them worse, will this create slot battles?

Also, some Dapps would like to see the max block size increase, the addition of Peras might limit this. Not sure what statement _can_ be made about this fact, but it might be worth adding a section on this balance for completeness?

Reply:
The idea was that the peras certificate would be significantly smaller than the block size so that block producers would fill in the remaining bytes and just propagate the block normally, so there should not be an impact on block diffusion, only on tx inclusion in case this happens while the mempools are congested. 

Note that inclusion of certificates in blocks will be an extremely rare event. In the worst case, assuming nodes immediately enter cooldown when they are back from it, this means 1 Peras certificate roughly every 3000 blocks. In normal operations, this could be much rarer as we expect quorum will be reached on every round.